pkg/keychain/keychain.go

```diff
@@ -0,0 +1,321 @@
+		// mainnet requires ledger usage
+		if keyName != "" || useEwoq {
+			return nil, ErrStoredKeyOrEwoqOnMainnet
+		}
+		useLedger = true
+	}
+
+	// will use default local keychain if simulating public network opeations on local
+	if os.Getenv(constants.SimulatePublicNetwork) != "" {
+		network = models.LocalNetwork
+	}
+
+	// get keychain accessor
+	return GetKeychain(app, useEwoq, useLedger, ledgerAddresses, keyName, network, requiredFunds)
+}
+
+func GetKeychain(
+	app *application.Avalanche,
+	useEwoq bool,
+	useLedger bool,
+	ledgerAddresses []string,
+	keyName string,
+	network models.Network,
+	requiredFunds uint64,
+) (*Keychain, error) {
+	// get keychain accessor
+	if useLedger {
+		ledgerDevice, err := ledger.New()
+		if err != nil {
+			return nil, err
+		}
+		// always have index 0, for change
+		ledgerIndices := []uint32{0}
+		if requiredFunds > 0 {
+			ledgerIndicesAux, err := searchForFundedLedgerIndices(network, ledgerDevice, requiredFunds)
+			if err != nil {
+				return nil, err
+			}
+			ledgerIndices = append(ledgerIndices, ledgerIndicesAux...)
+		}
+		if len(ledgerAddresses) > 0 {
+			ledgerIndicesAux, err := getLedgerIndices(ledgerDevice, ledgerAddresses)
+			if err != nil {
+				return nil, err
+			}
+			ledgerIndices = append(ledgerIndices, ledgerIndicesAux...)
+		}
+		ledgerIndicesSet := set.Set[uint32]{}
+		ledgerIndicesSet.Add(ledgerIndices...)
+		ledgerIndices = ledgerIndicesSet.List()
+		sortUint32(ledgerIndices)
+		if err := showLedgerAddresses(network, ledgerDevice, ledgerIndices); err != nil {
+			return nil, err
+		}
+		kc, err := keychain.NewLedgerKeychainFromIndices(ledgerDevice, ledgerIndices)
+		if err != nil {
+			return nil, err
+		}
+		return NewKeychain(network, kc, ledgerDevice, ledgerIndices), nil
+	}
+	if useEwoq {
+		sf, err := key.LoadEwoq(network.ID)
+		if err != nil {
+			return nil, err
+		}
+		kc := sf.KeyChain()
+		return NewKeychain(network, kc, nil, nil), nil
+	}
+	sf, err := key.LoadSoft(network.ID, app.GetKeyPath(keyName))
+	if err != nil {
+		return nil, err
+	}
+	kc := sf.KeyChain()
+	return NewKeychain(network, kc, nil, nil), nil
+}
+
+func getLedgerIndices(ledgerDevice keychain.Ledger, addressesStr []string) ([]uint32, error) {
+	addresses, err := address.ParseToIDs(addressesStr)
+	if err != nil {
+		return []uint32{}, fmt.Errorf("failure parsing ledger addresses: %w", err)
+	}
+	// maps the indices of addresses to their corresponding ledger indices
+	indexMap := map[int]uint32{}
+	// for all ledger indices to search for, find if the ledger address belongs to the input
+	// addresses and, if so, add the index pair to indexMap, breaking the loop if
+	// all addresses were found
+	for ledgerIndex := uint32(0); ledgerIndex < numLedgerIndicesToSearch; ledgerIndex++ {
+		ledgerAddress, err := ledgerDevice.Addresses([]uint32{ledgerIndex})
+		if err != nil {
+			return []uint32{}, err
+		}
+		for addressesIndex, addr := range addresses {
+			if addr == ledgerAddress[0] {
```

Comment:
`if len(ledgerAddress) && addr == ledgerAddress[0]` to make sure ledgerAddress is not empty to avoid crash